lib/tpm2.c

```diff
@@ -336,15 +336,75 @@ tool_rc tpm2_context_load(ESYS_CONTEXT *esys_context,
     return tool_rc_success;
 }
 
-tool_rc tpm2_flush_context(ESYS_CONTEXT *esys_context, ESYS_TR flush_handle) {
+tool_rc tpm2_flush_context(ESYS_CONTEXT *esys_context, ESYS_TR flush_handle,
+    TPM2B_DIGEST *cp_hash, TPMI_ALG_HASH parameter_hash_algorithm) {
+
+   tool_rc rc = tool_rc_success;
+   TSS2_RC rval = TSS2_RC_SUCCESS;
+    if (cp_hash && cp_hash->size) {
+        /*
+         * Need sys_context to be able to calculate CpHash
+         */
+        TSS2_SYS_CONTEXT *sys_context = 0;
+        rc = tpm2_getsapicontext(esys_context, &sys_context);
+        if(rc != tool_rc_success) {
+            LOG_ERR("Failed to acquire Tss2_Sys_FlushContext_Prepare SAPI context.");
+            return rc;
+        }
 
-    TSS2_RC rval = Esys_FlushContext(esys_context, flush_handle);
+        TSS2_RC rval = Tss2_Sys_FlushContext_Prepare(
+        sys_context, flush_handle);
+        if (rval != TPM2_RC_SUCCESS) {
+            LOG_PERR(Tss2_Sys_FlushContext_Prepare, rval);
+            return tool_rc_general_error;
+        }
+
+        /*
+         * There is a bug in SAPI where in the flush handle is placed in the
+         * handle area instead of the parameter area.
+         * Ref: https://github.com/tpm2-software/tpm2-tss/issues/2382
+         *
+         * We determine this scenario by reading the parameter size in the
+         * cpBuffer which is returned as zero due to the bug. 
+         *
+         * When calculating the cpHash, the workaround for this scenario is to
+         * provide the flush handle name.
+         */
+        const uint8_t *command_parameters;
+        size_t command_parameters_size;
+        rval = Tss2_Sys_GetCpBuffer(sys_context, &command_parameters_size,
+            &command_parameters);
+        if (rval != TPM2_RC_SUCCESS) {
+            LOG_PERR(Tss2_Sys_GetCpBuffer, rval);
+            return tool_rc_general_error;
+        }
+
+        TPM2B_NAME *name1 = 0;
+        bool is_param_size_zero = (command_parameters_size == 0);
+        if (is_param_size_zero) {
```

Comment:
Their is no need to assign to a bool, a simple `if (command_parameters_size == 0)` or `(!command_parameters_size)` is sufficient.